### PR TITLE
Add Kinesis Source

### PR DESCRIPTION
**Pull request by zsxwing**
This PR adds the following changes based on https://github.com/tdas/spark/tree/streaming-df-kinesis

- Clean up old blocks when getNextBatch is called.
- Add kinesis method to DataFrameReader like spark-avro here: https://github.com/databricks/spark-avro/blob/master/src/main/scala/com/databricks/spark/avro/package.scala
- Unit tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockId.scala [hunk 1]
```diff
@@ -85,6 +85,11 @@ case class StreamBlockId(streamId: Int, uniqueId: Long) extends BlockId {
   override def name: String = "input-" + streamId + "-" + uniqueId
 }
 
+@DeveloperApi
+case class StreamSourceBlockId(source: String, uniqueId: Long) extends BlockId {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockId.scala]: Because `StreamBlockId` uses `receiverId` as `streamId` and KinesisSource doesn't know anything about receivers, it cannot generate a unique streamId. So I just added a new BlockId type.

**Comment by zsxwing** [comment]
Steps to run unit tests:
- Set up aws credentials environment: AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
- `export ENABLE_KINESIS_TESTS=1`

File: core/src/main/scala/org/apache/spark/storage/BlockId.scala (hunk 1 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockId.scala]: This is the thing. I am not sure its a good idea to introduce a new "BlockId" definition in core.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  }
+
+  /**
+    * Get the records starting from using a Kinesis shard iterator (which is a progress handle
+    * to get records from Kinesis), and get the next shard iterator for next consumption.
+    */
+  private def getRecordsAndNextKinesisIterator(
+      shardIterator: String): (Seq[Record], String) = {
+    val getRecordsRequest = new GetRecordsRequest().withShardIterator(shardIterator)
+    getRecordsRequest.setRequestCredentials(credentials)
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: Is there no `withRequestCredentials`?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  override def shortName(): String = "kinesis"
+
+  override def createSource(
+      sqlContext: SQLContext,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+    val regionName = caseInsensitiveOptions.getOrElse("regionName", {
+      throw new IllegalArgumentException("regionName is not specified")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Isnt it better to say "Option `regionName` not specified"?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      throw new IllegalArgumentException("endpointUrl is not specified")
+    })
+    val streamNames = caseInsensitiveOptions.getOrElse("streamNames", {
+      throw new IllegalArgumentException("streamNames is not specified")
+    }).split(',').toSet
+    val initialPosInStream =
+      caseInsensitiveOptions.getOrElse("initialPosInStream", "LATEST") match {
+        case "LATEST" => InitialPositionInStream.LATEST
+        case "TRIM_HORIZON" => InitialPositionInStream.TRIM_HORIZON
+        case pos => throw new IllegalArgumentException(s"unknown initialPosInStream: $pos")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: "Unknown value of option `initialPosInStream`: $pos"

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      }
+    val awsCredentialsOption =
+      for (accessKeyId <- caseInsensitiveOptions.get("AWS_ACCESS_KEY_ID");
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: We should throw error if only one of them is specified. Either both should be specified, or none at all.

File: core/src/main/scala/org/apache/spark/storage/BlockId.scala (hunk 1 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockId.scala]: I think a better idea might be to modify the existing StreamBlockId to accept an optional string (which by default would be "input"). Something like 
```
case class StreamBlockId(context: String, streamId: Int, uniqueId: Long) extends BlockId {
     def this(streamId: Int, uniqueId: Long) = this("input", streamId, uniqueId)
     override def name: String = s"$context-$streamId-$uniqueId"
}
```

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+  private val nextId = new AtomicLong(0)
+
+  def nextBlockId: StreamSourceBlockId = StreamSourceBlockId("kinesis", nextId.getAndIncrement)
+}
+
+/**
+ * However, this class runs in the driver so could be a bottleneck.
+ */
+private[kinesis] class KinesisDataFetcher(
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: Would be a good idea to split this into a different file.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 7]
```diff
@@ -334,7 +339,7 @@ private[kinesis] class KinesisReceiver[T](
      * The data addition, block generation, and calls to onAddData and onGenerateBlock
      * are all synchronized through the same lock.
      */
-    def onAddData(data: Any, metadata: Any): Unit = {
+    override def onAddData(data: Any, metadata: Any): Unit = {
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: These changes are unrelated.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        }
+      }
+
+      val nonZeroSigns = comparisons.filter(_ != 0).toSet
+      nonZeroSigns.size match {
+        case 0 => 0 // if both empty or only 0s
+        case 1 => nonZeroSigns.head // if there are only (0s and 1s) or (0s and -1s)
+        case _ => // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $that")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: "Invalid comparison between KinesisSource offsets: this: $this\nthat: $that"

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala (hunk 2 above)
Comment (zsxwing) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: no...
